src/p4est_io.h

```diff
@@ -68,6 +108,398 @@ p4est_t            *p4est_inflate (sc_MPI_Comm mpicomm,
+ * these two types of blocks.
+ * Every data block is preceded by 64 bytes block header written
+ * by p4est. These 64 bytes are again written to the file as string* and can
+ * be read using a text editor.
+ *
+ * Block Header (64 bytes):
+ * One byte block type specific character (H for a header block and F for
+ * a data array), 1 byte space and 13 bytes size in number of bytes for a
+ * header block and data size per element in byte for a data array block
+ * and one trailing byte new line char.
+ * 47 bytes user-defined string* and 1 byte new line char.
+ *
+ * The structure of p4est and p8est data files differs only by the magic number.
+ *
+ * The p4est metadata of a p4est data file can be accessed by \ref p4est_file_info().
+ */
+
+/** Opaque context used for writing a p4est data file. */
+typedef struct p4est_file_context p4est_file_context_t;
+
+/** Begin writing file header and saving data blocks into a parallel file.
+ *
+ * This function creates a new file or overwrites an existing one.
+ * It is collective and creates the file on a parallel file system.
+ * It takes an (optional) pointer to write a header of given size.
+ * This function leaves the file open if MPI I/O is available.
+ * It is necessary to call \ref
+ * p4est_file_close (possibly after writing one or more data sets).
+ * The file is opened in a write-only mode.
+ *
+ * We add some basic metadata to the file.
+ * The file written contains the file header and data sets
+ * as specified by the open/write functions called.
+ * The file header consists of the metadata specified by p4est.
+ *
+ * It is the application's responsibility to write sufficient header
+ * information (cf. \ref p4est_file_write_header) to determine the number and
+ * size of the data sets if such information is not recorded and maintained
+ * externally.
+ * However, p4est makes some metadata accessible via
+ * \ref p4est_file_info.
+ *
+ * This function does not abort on MPI I/O errors but returns NULL.
+ *
+ * \param [in] p4est          Valid forest.
+ * \param [in] filename       Path to parallel file that is to be created.
+ * \param [in] user_string    A user string that is written to the file header.
+ *                            Only 15 bytes without null-termination are
+ *                            written to the file. If the user gives less
+ *                            bytes the user_string in the file header is padded
+ *                            by spaces.
+ * \param [out] errcode       An errcode that can be interpreted by \ref
+ *                            p4est_file_error_string and
+ *                            \ref p4est_file_error_class.
+ * \return                    Newly allocated context to continue writing
+ *                            and eventually closing the file. NULL in
+ *                            case of error.
+ */
+p4est_file_context_t *p4est_file_open_create
+  (p4est_t * p4est, const char *filename, const char user_string[15],
```

Comment:
The number 15 differs from the above `USER_STRING_BYTES` of 48. Does it make sense to unify these? Also, shall we use a `#define`d value and not the bare number 15 (what is our usual convention)? With the `_ext` version, we provide user_string as `char *`, again, does it make sense to unify?

Reply:
Please always force the user to input `NUL`-terminated strings, and check that the input length is < 16 (in this case).

Reply:
> The number 15 differs from the above `USER_STRING_BYTES` of 48. Does it make sense to unify these? Also, shall we use a `#define`d value and not the bare number 15 (what is our usual convention)?

Yes, I unified the number of bytes of the user string and used the macro for the number of bytes.

>  With the _ext version, we provide user_string as char *, again, does it make sense to unify?

Yes, I used also the macro for number of bytes of the user string at this place.

Reply:
> Please always force the user to input `NUL`-terminated strings, and check that the input length is < 16 (in this case).

This is now also done including assertions using `strlen`